Convert session bar into a pulldown menu

foxbunny opened the pull request:
- Add pulldown menu
- Rename session bar into session menu
- Organize protection info, locale and tz logic into dedicated classes
- Move logic from session menu into Python and make it reusable
- Modify change_lang and change_tz controllers to support plain forms
- Remove jQuery from locale and tz forms
- Add ind-toggle-trigger custom elemenet
- Add new protection indicator partial and template global for use in page headers
- Adjust header height to (roughly) match the RB module

Fixes #5966

foxbunny commented on indico/modules/core/controllers.py:
```diff
@@ -138,6 +139,8 @@
             session.user.settings.set('force_timezone', mode != 'local')
 
+        if redirect_url:
+            return redirect(redirect_url, 303)
```
@ThiefMaster This is the issue that we talked about before. If my memory serves me, you said it was fine, but let me know if we should address it. I don't have any fantastic ideas, though.

ThiefMaster replied:
Since it's a POST request I think it's very hard to abuse it as an open redirector. However, assuming this is always going to be a relative URL I recommend a check like this just in case:

```python
if redirect_url.startswith('//') or not redirect_url.startswith('/'):
    redirect_url = '/'
```

I think this prevents any non-relative URLs (since you cannot use `//evil.com/...` nor `https://evil.com`).

ThiefMaster commented on indico/util/locales.py:
```diff
@@ -0,0 +1,63 @@
+        return self.forced_context_locale is not None
+
+    @cached_property
+    def has_restircted_choice(self):
```
typo -> restricted

ThiefMaster commented on indico/web/templates/_session_menu.html:
```diff
@@ -0,0 +1,220 @@
+    {% set title %}{% trans %}Timezone{% endtrans %}{% endset %}
+    {% call _settings_section('tz-selector-controls', 'tz-selector', title) %}
+        {% if timezone_data.has_no_choice %}
+            <p class="tz-locked-note">{% trans active_tz=timezone_data.active_tz %}The active timezone is {{ active_tz }} and it cannot be changed{% endtrans %}.</p>
```
Maybe this can be worded a bit more nicely?

"Management areas always use the timezone of the event/category" or something like that...

foxbunny replied:
Is that the only place where timezone is locked?

ThiefMaster commented on indico/util/protection_info.py:
```diff
@@ -0,0 +1,60 @@
+
+    @cached_property
+    def disclaimer(self):
+        if self.mode == 'network':
```
`self.mode` does not exist

foxbunny replied:
Ah damn, yeah, I'll fix that when I rename "protection_mode".

ThiefMaster commented on indico/util/protection_info.py:
```diff
@@ -0,0 +1,60 @@
+        return self.object.is_protected
+
+    @cached_property
+    def access_restirctions(self):
```
typo -> restrictions

ThiefMaster commented on indico/util/protection_info.py:
```diff
@@ -0,0 +1,60 @@
+        return sorted(o.name for o in self.access_restirctions if o.principal_type.name == 'network')
+
+    @cached_property
+    def protection_mode(self):
```
Can we use a different name for this? `protection_mode` is already used elsewhere for inheriting/public/protected and it might be a bit misleading to use the same name for something else. How about `protection_level` instead?

ThiefMaster commented on indico/util/timezones.py:
```diff
@@ -0,0 +1,52 @@
+    @cached_property
+    def all_tzs(self):
+        if self.active_tz not in common_timezones_set:
+            return [*list(common_timezones), self.active_tz]
```
is the `list()` actually needed? usually star-unpacking works without casting to a list first

ThiefMaster commented:
Could we keep the current timezone visible? Especially in management areas where it's locked to the event's timezone it's useful, but I think on event/category pages it's also nice to have it so you immediately know what timezone is being used on that page.

foxbunny commented on indico/web/templates/_session_menu.html:
```diff
@@ -0,0 +1,220 @@
+                <form action="{{ url_for('core.change_lang') }}" method="post">
+                    <input type="hidden" name="url" value="{{ request.url }}">
+                    <input type="hidden" name="csrf_token" value="{{ session.csrf_token }}">
+                    <ul id="language-list">
```
If I figure out a good way to take care of that I'll add it in a separate PR. One thing that We could change in this PR is maybe add a maximum *width*, though. That will reduce the area of the empty space, and frankly it doesn't need to be that wide to accommodate the UI elements.

foxbunny commented on indico/util/locales.py:
```diff
@@ -0,0 +1,63 @@
+
+    def get_locale_display(self, locale):
+        (name, territory, is_ambiguous) = self.get_locale_info(locale)
+        if is_ambiguous:
+            return f'{name} ({territory})'
```
That actually means "Portugese" *is* ambiguous, no?

foxbunny replied:
That is, it's a data issue, not UI issue.

foxbunny replied:
So you currently use `is_ambiguous` to mean "We have more than 1 translation for languages with the same name", but it actually should mean "We have more than 1 language with the same name (regardless of whether there are translation for it)". It's a domain issue, not a presentation issue.

ThiefMaster replied:
ah, i didn't realize `is_ambiguous` isn't new but something we already had. Now that I noticed it and checked how we're using it I remember why it's there:

When showing the current language I wanted to avoid being overly verbose except when it's needed, but when showing the languages to choose from I'm always showing the full name with the territory included. That way you know what you get when changing the language, but once you are using the language we prefer to show just the short name - unless Indico has more than one version of that language.

foxbunny replied:
Ok

foxbunny replied:
In that case, I think you don't really need the `is_ambiguous` flag at all?

foxbunny commented on indico/web/templates/_session_menu.html:
```diff
@@ -0,0 +1,220 @@
+                </p>
+                {% if protection_data.disclaimer %}
+                    <p class="protection-disclaimer">
+                        {{ protection_data.disclaimer|sanitize_html|trim }}
```
Hm, yeah. I didn't notice that case. I'll figure something out.

foxbunny commented on indico/web/templates/header.html:
```diff
@@ -5,25 +5,34 @@
+            {{ render_protection_indicator(protected_object) }}
+            <ind-toggle-trigger>
+                <button class="header-menu-trigger" aria-controls="session-menu">
+                    {% trans %}Menu{% endtrans %}
```
Your call. I have no strong opinions on that. I think both would be pretty clear.

ThiefMaster replied:
i asked the team and so far the preference goes for hamburger icon (we already use that in room booking)

foxbunny commented:
> Could we keep the current timezone visible in the header? Especially in management areas where it's locked to the event's timezone it's useful, but I think on event/category pages it's also nice to have it so you immediately know what timezone is being used on that page.

Here's what I'm thinking. Showing the TZ in the header is going to make it a big busy, and we're basically back to where we were. There's space on the right-hand side of the global menu area, though, where I can place the time zone indicator. I can place a "lock" icon or something on it to indicate that it's locked.

Something like this:

<img width="296" alt="image" src="https://github.com/indico/indico/assets/65413/75d0ffee-21ed-454b-ba45-1e7a669b7c65">

foxbunny commented on indico/web/templates/_session_menu.html:
```diff
@@ -0,0 +1,220 @@
+                        {% endtrans %}
+                    {% endif %}
+                </p>
+                {% if protection_data.disclaimer %}
```
I was thinking about it. Having two triggers to do the same thing is quite awkward from the accessibility point of view, but maybe I can do some voodoo with links instead.

foxbunny replied:
Another option is to take the message out of the menu and add it as a toggle tip (info icon) to the indicator.

ThiefMaster replied:
toggletip doesn't sound bad for it. it's pretty similar to what we have right now...

ThiefMaster commented:
I like the idea of putting the timezone there (using the full name though, e.g. `🕰 Europe/Zurich`). I'll also let the other guys in the team comment on it though what they think.

foxbunny commented:
@ThiefMaster I've made the changes as per our discussions. I've marked our conversation as resolved where I thought the conclusions were clear. I've left some of them open where I would like to solicit further feedback, so please check them when you can.

ThiefMaster commented:
Could we somehow keep showing the logged-in user? **At least** the profile picture?

For developers testing things no longer immediately seeing whether they are logged in (and with which account) will be terrible, but also as a user it's nice to see if you are logged-in or not. Thinking of cases like public (meeting room) computers. You want to quickly see whether you still need to logout or not.

Maybe a label similar to what we have for the protection indicator? Instead of the icon the user's profile picture could be used, and the text could be the initial and last name like we have right now.

foxbunny commented:
I'll figure something out once I'm done with the event page stuff and the settings dropdowns.

ThiefMaster commented:
I'm also going to push a few small fixes here which I found while testing this

ThiefMaster commented on indico/web/client/styles/partials/_protection_indicator.scss:
```diff
@@ -0,0 +1,60 @@
+@use 'partials/icons';
+@use 'design_system';
+
+#protection-indicator {
```
With`align-items: center` this looks much better since the icon is nicely centered:

old:
![image](https://github.com/indico/indico/assets/179599/0bc6a46d-4723-4115-a643-8ceb25009f70)
new:
![image](https://github.com/indico/indico/assets/179599/32c922e9-31b3-4f4a-9eef-730fe366fe6b)

ThiefMaster commented:
Could keep two different colors for 'properly protected' and 'network grants access'?

Maybe also showing the network name in the label as well, but I don't have a super strong opinion on that, would need to see if it looks good or not...

ThiefMaster commented:
```css
body {
  overflow-y: scroll;
}
```

would probably be good. Otherwise opening the menu looks weird when there was no scrollbar before. Better to always have a disabled scroll bar on the page if no scrolling is needed...

tomasr8 commented:
> > Could we keep the current timezone visible in the header? Especially in management areas where it's locked to the event's timezone it's useful, but I think on event/category pages it's also nice to have it so you immediately know what timezone is being used on that page.
> 
> Here's what I'm thinking. Showing the TZ in the header is going to make it a big busy, and we're basically back to where we were. There's space on the right-hand side of the global menu area, though, where I can place the time zone indicator. I can place a "lock" icon or something on it to indicate that it's locked.
> 
> Something like this:
> <img alt="image" width="296" src="https://private-user-images.githubusercontent.com/65413/296712457-75d0ffee-21ed-454b-ba45-1e7a669b7c65.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU1OTE4NDksIm5iZiI6MTcwNTU5MTU0OSwicGF0aCI6Ii82NTQxMy8yOTY3MTI0NTctNzVkMGZmZWUtMjFlZC00NTRiLWJhNDUtMWU3YTY2OWI3YzY1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE4VDE1MjU0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNkZGYzZGEyNDE0ZTc2MTA2MmMzNzE5NDgzZTBmZTYxNTI0NzYwMmQyMDU4MjIzZjcwNmJmZGI1NjhmZTg5N2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0r4XKIGnfqooVpwoi9aZtBLPxRY_M6S9kzmaDKGNfJ8">

Not sure how well that's gonna like look on meeting pages:
![obrazek](https://github.com/indico/indico/assets/8739637/48f640eb-ad2f-48b7-96f1-c15aee1652a8)

For me, I think I'd be fine having it on the same line as the protection info

tomasr8 commented:
btw, I like the color for the 'protected' chip for meeting pages, but I think the color doesn't mesh very well with a blue background:
![obrazek](https://github.com/indico/indico/assets/8739637/fff95334-6923-446d-81ac-177b2b6f6de4)
![obrazek](https://github.com/indico/indico/assets/8739637/8f7a6baa-9bca-476b-a389-cfe1d0546e32)

tomasr8 commented:
> Could we somehow keep showing the logged-in user? **At least** the profile picture?

Agreed, with the new session menu it took me a while to figure out that I wasn't logged in..

ThiefMaster commented:
> Not sure how well that's gonna like look on meeting pages

I think on meeting pages the timezone in the header is not *that* important since we show it next to the event time anyway. That's why I didn't suggest adding it there as well when I noticed earlier today that it's not shown there.

But if someone has a nice idea on how to show it there nicely, I wouldn't be opposed to having it either.

tomasr8 commented:
> > Not sure how well that's gonna like look on meeting pages
> 
> I think on meeting pages the timezone in the header is not _that_ important since we show it next to the event time anyway. That's why I didn't suggest adding it there as well when I noticed earlier today that it's not shown there.
> 
> But if someone has a nice idea on how to show it there nicely, I wouldn't be opposed to having it either.

True, then I think we can leave it out for those pages

duartegalvao commented:
The animation for opening and closing the header menu seems a bit off - looks a bit weird to me and doesn't really fit with the Indico style tbh. Could you do maybe a short fade in and out? If not, I think it'd look fine even with no animation at all.

foxbunny commented:
> but I think the color doesn't mesh very well with a blue background

My intent here is to use color as part of the information architecture, so having a single color that means "protected" would be great. I sort of reused the same color that was used before (for the shield) after adjusting it slightly for color contrast.

Now, since the chosen color doesn't play nice with the background, I have two solutions. One is to change the color to a brighter shade or red (orange doesn't look so good when made brighter, and won't work on a light background). Alternatively, we can do something like this, which I think looks pretty cool:

<img width="961" alt="image" src="https://github.com/indico/indico/assets/65413/7cd50354-702d-4b3b-bd96-f9db280b829b">

foxbunny commented:
> The animation for opening and closing the header menu seems a bit off - looks a bit weird to me and doesn't really fit with the Indico style tbh. 

Can you point me to an existing animation for opening larger surfaces? I'll try to match it.

foxbunny commented:
> Maybe also showing the network name in the label as well

I'd avoid that since (1) it's shown in the toggle tip, (2) there could be multiple network names and/or network names could be long, so it would make the label quite large, breaking the layout.

foxbunny commented:
Rebased, and addressed most of the comments. The only remaining thing is the animation mentioned by @duartegalvao. I need some clarity regarding the desired style. An example would be helpful.

duartegalvao commented:
> @duartegalvao Can you point me to an existing animation for opening larger surfaces? I'll try to match it.

We don't really have any such type of animation. The closest thing I can think of to "opening larger surfaces" are modals, which don't have any animation, which is honestly what I would do in this case

foxbunny commented:
> The closest thing I can think of to "opening larger surfaces" are modals, which don't have any animation, which is honestly what I would do in this case

From UX point of view, modal is a different use case because it does not cause a layout shift (well, shouldn't anyway 😂). The pull down menu pushes the the content below, so without animation, it causes confusion. It's similar in behavior to an accordion (where I would also use an animation).

Anyway, let me know if you insist on removing the animation. I think it'll be worse for UX, but not the hill I'll die on.

foxbunny commented:
FWIW There's an example of an expanding surface in the timetable view on the conference page. It uses a rather slow version of the animation I applied to the menu. It appears to be using a linear easing function, and not counting the glitch when I expand it, it looks like it's doing about the same. The only difference is that it doesn't close with an animation, which I think is a bug. Also, the animation speed seems more like a mistake than a feature. It could definitely be a lot faster.

ThiefMaster commented:
I would not use anything from the (ultra-)legacy conference timetable JS (it's from over 10 years ago!) as a particularly good example on how something should be done ;)

foxbunny commented:
> I would not use anything from the (ultra-)legacy conference timetable JS

I'm just trying to get an idea of what "Indico style animation" is like.

duartegalvao commented:
> > The closest thing I can think of to "opening larger surfaces" are modals, which don't have any animation, which is honestly what I would do in this case
> 
> From UX point of view, modal is a different use case because it does not cause a layout shift (well, shouldn't anyway 😂). The pull down menu pushes the the content below, so without animation, it causes confusion. It's similar in behavior to an accordion (where I would also use an animation).
> 
> Anyway, let me know if you insist on removing the animation. I think it'll be worse for UX, but not the hill I'll die on.

I was checking the PR on my instance and yeah, I agree removing the animation would be worse.

I was noticing what you did with the other dropdowns (create event, services) and they look really good! Do you think something similar would be possible with this menu? This means that the surface would be overlaid on top of the page content instead of pushing it away, but IMO it would be better...

foxbunny commented:
> If this is too complicated or you think it would be bad UX, I'm fine with merging the PR as is. It looks really nice overall

I'm inclined to say that we move forward with the merge anyway for now, and consider further improvements later. I don't expect it to be a major improvement in the UX even if we find a simple way to mimic the dropdown animation style, plus I have a huge backlog of issues that are waiting for various PRs to get merged. 😉

ThiefMaster commented:
I'd be OK with merging it into `a11y` since all the technical improvements are indeed good to have available for the other PRs.

However, I think before `a11y` goes into `master`, we should try to polish the UX/looknfeel as needed so the end users get something "pretty".

And maybe before merging this one into a11y I'll move the two commits already merged into that branch to master since I believe there are no open issues with those anymore.
